lsp-dired.el

```diff
@@ -0,0 +1,161 @@
+    (let (buffer-read-only)
+      (save-excursion
+        (goto-char pos)
+        (forward-line (if lsp-dired--ranger-adjust 1 2))
+        (cl-block :file
+          (while (not (eobp))
+            (if (dired-move-to-filename nil)
+                (let ((file (dired-get-filename nil t))
+                      (bol (progn
+                             (point-at-bol)
+                             (search-forward-regexp "^[[:space:]]*" (line-end-position) t)
+                             (point))))
+                  (-doto (make-overlay bol (point-at-eol))
+                    (overlay-put 'lsp-color t)
+                    (overlay-put 'evaporate t)
+                    (overlay-put 'face (lsp-dired--face-for-path file) )))
+              (cl-return-from :file nil))
+            (forward-line 1)))))))
+
+(defface lsp-dired-path-face '((t :inherit font-lock-string-face))
+  "Face used for breadcrumb paths on headerline."
+  :group 'lsp-faces)
+
+(defface lsp-dired-path-error-face
+  '((t :underline (:style wave :color "Red1")))
+  "Face used for breadcrumb paths on headerline when there is an error under that path"
+  :group 'lsp-faces)
+
+(defface lsp-dired-path-warning-face
+  '((t :underline (:style wave :color "Yellow")))
+  "Face used for breadcrumb paths on headerline when there is an warning under that path"
+  :group 'lsp-faces)
+
+(defface lsp-dired-path-info-face
+  '((t :underline (:style wave :color "Green")))
+  "Face used for breadcrumb paths on headerline when there is an info under that path"
+  :group 'lsp-faces)
+
+(defface lsp-dired-path-hint-face
+  '((t :underline (:style wave :color "Green")))
+  "Face used for breadcrumb paths on headerline when there is an hint under that path"
+  :group 'lsp-faces)
+
+(defun lsp-dired--face-for-path (dir)
+  "Calculate the face for DIR."
+  (when-let ((diags (lsp-diagnostics-stats-for (directory-file-name dir))))
+    (cl-labels ((check-severity
+                 (severity)
+                 (not (zerop (aref diags severity)))))
+      (cond
+       ((check-severity lsp/diagnostic-severity-error)
+        'lsp-dired-path-error-face)
+       ((check-severity lsp/diagnostic-severity-warning)
+        'lsp-dired-path-warning-face)
+       ((check-severity lsp/diagnostic-severity-information)
+        'lsp-dired-path-info-face)
+       ((check-severity lsp/diagnostic-severity-hint)
+        'lsp-dired-path-hint-face)))))
+
+(defun lsp-dired--insert-subdir-advice (&rest args)
+  "Advice to dired & dired+ insert-subdir commands.
+Will add icons for the subdir in the `car' of ARGS."
+  (let* ((path (car args))
+         (pos (cdr (assoc path dired-subdir-alist))))
+    (when pos
+      (lsp-dired--insert-for-subdir path pos))))
+
+(defun lsp-dired--kill-subdir-advice (&rest _args)
+  "Advice to dired kill-subdir commands.
+Will remove the killed subdir from `lsp-dired--covered-subdirs'."
+  (setf lsp-dired--covered-subdirs (delete (dired-current-directory)
```

Comment:
You could use `cl-callf2` here, although it is somewhat obscure:
```emacs-lisp
(cl-callf2 delete (dired-current-directory) lsp-dired--covered-subdirs)
```

Reply:
To me, it seems like using callf2 will be beneficial when the `place` is more complex while setq/setf seems to be good enough for symbols.